aten/src/ATen/functorch/BatchRulesScatterOps.cpp

```diff
@@ -1056,6 +1056,150 @@ std::tuple<Tensor,optional<int64_t>> masked_fill_scalar_batch_rule(
   return std::make_tuple(result, 0);
 }
 
+std::tuple<Tensor,optional<int64_t>> index_fill_int_scalar_batch_rule_impl(
+    Tensor & self, optional<int64_t> self_bdim,
+    int64_t dim,
+    const Tensor & index, optional<int64_t> index_bdim,
+    const Scalar & value,
+    const bool inplace) {
+
+    Tensor self_ = moveBatchDimToFront(self, self_bdim);
+    Tensor index_ = moveBatchDimToFront(index, index_bdim);
+    dim = maybe_wrap_dim(dim, self_.dim());
+
+    if (inplace && !self_bdim.has_value()) {
+      vmapIncompatibleInplaceError("index_fill_");
+    }
+
+    if (!index_bdim) {
+      self_.unsqueeze_(-1).index_fill_(dim + 1, index_, value).squeeze_(-1);
+      return std::make_tuple(self_, 0);
+    }
+
+    auto batch_size = get_bdim_size2(self, self_bdim, index, index_bdim);
+    self_ = ensure_has_bdim(self_, self_bdim.has_value(), batch_size);
+    index_ = ensure_has_bdim(index_, index_bdim.has_value(), batch_size);
+
+    if (inplace) {
+      // Do for-loop for in-place because we cannot reshape
+      // `self_` having an incompatible stride without copying
+      for (const auto i : c10::irange(0, batch_size)) {
+        const auto& self_slice = self_.select(0, i);
+        const auto& index_slice = index_.select(0, i);
+        self_slice.index_fill_(
+          dim,
+          index_slice,
+          value
+        );
+      }
+    } else {
+      self_ = self_bdim.has_value() ? self_ : self_.clone();
+
+      if (self.dim() > 1){
+        if (dim == 0) {
+          // The dim to which we're applying is batch dim which will be packed into the next dim for batched calculution.
+          // We add offsets to index to conform with the element positions of the reshaped self_.
+          auto index_offset = at::arange(
+            batch_size,
+            at::TensorOptions().dtype(index_.scalar_type()).device(index_.device())
+          );
+          index_ = index_.add(index_offset.unsqueeze(-1), self_.size(1));
+        }
+
+        index_ = reshape_dim_into(0, 0, index_);
+        self_ = reshape_dim_into(0, 0, self_);
+        self_.index_fill_(dim, index_, value);
+        self_ = reshape_dim_outof(0, batch_size, self_);
+
+      } else {
+        // If self.dim() is 0 or 1, the batch dim is certainly 0, and we must apply batched indices to each row.
+        index_ = reshape_dim_into(0, 0, index_);
+        self_.unsqueeze_(-1).index_fill_(dim + 1, index_, value).squeeze_(-1);
```

Comment:
Is this case just to improve performance? If so, I'd prefer to remove it so that all of our code goes through the above case: it makes the code a bit simpler to reason about if there's just a single case for out-of-place

Reply:
No it's not just for performance improvement (or maybe I'm missing something). In case of self.dim() == 0 or 1 because the only dim is the batch dim, we need to add another dim at the last in order to apply batched index to it.

For example: 

If `self` is `[1, 2]` and a batched `index`, where the batch dim is 0, is `[[0], [0]]`, we should fill `value` to self[0] and self[1] instead of only self[0].